Fix PL2Bat (ERRORLEVEL and environment handling), tests, and AppVeyor CI

rivy opened the pull request:
PR includes multiple fixes (further detailed in the commits):

- PL2Bat fixes for "echo", "ERRORLEVEL", and environment state handling
- Testing includes process exit status as well as expanded exit value testing
- AppVeyor now correctly tests for perl versions back to v5.8.9

Leont commented on t/make_executable.t:
```diff
@@ -2,32 +2,50 @@
 use Cwd qw/cwd/;
 
-plan($^O eq 'MSWin32' ? (tests => 7) : skip_all => 'Only usable on Windows');
+use IPC::System::Simple qw/ capture system EXIT_ANY /;
```
As this distribution is intended to be included in core, non-core dependencies are not appropraite.

rivy commented:
Ok.

I'm happy to change it... suggestions on a core alternative?

On 12 Feb 2018 03:54, "Leon Timmermans" <notifications@github.com> wrote:

> *@Leont* commented on this pull request.
> ------------------------------
>
> In t/make_executable.t
> <https://github.com/Perl-Toolchain-Gang/extutils-pl2bat/pull/3#discussion_r167505848>
> :
>
> >
>  use Config;
>  use Test::More;
>  use ExtUtils::PL2Bat;
>  use Cwd qw/cwd/;
>
> -plan($^O eq 'MSWin32' ? (tests => 7) : skip_all => 'Only usable on Windows');
> +use IPC::System::Simple qw/ capture system EXIT_ANY /;
>
> As this distribution is intended to be included in core, non-core
> dependencies are not appropraite.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Perl-Toolchain-Gang/extutils-pl2bat/pull/3#pullrequestreview-95733037>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAE5BFzS9Qh9xjZXdF84BCEV8i_4Bm_Uks5tUApqgaJpZM4SBIiq>
> .
>

rivy commented:
I refactored the tests to use only `qx/.../` command execution. I added some minor CMD shell hackery to enable testing the ERRORLEVEL from `qx` (which otherwise mangles the ERRORLEVEL/exit code down into a 7 bit return value in $CHILD_ERROR).

Let me know if that is an improvement and/or needs more modifications.

rivy commented:
@Leont , I fixed another couple of typos.

Is the current version now acceptable?

Leont commented:
> Is the current version now acceptable?

I'll discuss it on the [Perl Toolchain Summit](http://act.qa-hackathon.org/pts2018/)

rivy commented:
@Leont ... just checking back in; it's been over a year. In the meanwhile, I've added CI testing for windows and linux (via AppVeyor and Travis), both successfully testing back to perl v5.8.

Let me know if you need something else to complete this PR.

Leont commented:
I primarily need someone with more windows knowledge than me to do some reviewing. @haarg?

haarg commented:
`setlocal` requires `cmd.exe` and thus Windows NT.  It doesn't make sense to use it as well as trying to have other handling for `command.com`.  If we want to drop support for `command.com`, the entire script could be substantially simplified.

rivy commented:
Hmm, I believe I could make it work without `setlocal`. `%ERRORLEVEL%` would be clobbered by the script, but that seems reasonable. Otherwise, I'd just need to remove the `@echo off` line and add a lot of `@` prefixes to prevent echo'd lines.

rivy commented:
And, @haarg , thanks for the code review fly-by!

rivy commented:
Yes, I can eliminate `setlocal` and `echo off`.

I added a test to assure that there's no extraneous output and then refactored the BAT wrap without `setlocal` or `echo off`, using only leading `@` to suppress output (see <https://github.com/rivy/perl.ExtUtils-PL2Bat/commit/1a120d5562e5e9f6ccaa6796d241d1dc290d72d3> with AppVeyor testing at <https://ci.appveyor.com/project/rivy/perl-extutils-pl2bat/builds/22471711>).

If that seems like a better idea, I'll clean up the commits/messages and push it back onto this feature branch.

Leont commented:
> It doesn't make sense to use it as well as trying to have other handling for `command.com`. If we want to drop support for `command.com`, the entire script could be substantially simplified.

That sounds like a good idea. Though right now the "priority" is to get all three other forks of this codebase to use this thing first, so that improvements are shared.

Leont commented:
Why add the travis configuration if the tests will be skipped on non-windows?

rivy commented:
Good question.
I believe I put it in for completeness to make sure the build/test functionality works correctly under *nix, especially since `dzil` is not usable or testable on the AppVeyor Windows side. And, in looking at the Travis CI report, `dzil` the "t/00-compile.t" test passes (brought in by travis helpers?).

But it obviously doesn't need to test multiple versions. That's just an artifact of me putting my Travis CI config template in, instead of customizing it for this module. I should at least just cut it down to one or two versions, maybe just "stable".

Ultimately, if it's not useful information, I can cut that one commit back out.

rivy commented:
@Leont , what do I need to do to get this over the finish line? It's been on my PR list for over a year now.

Leont commented:
> what do I need to do to get this over the finish line? It's been on my PR list for over a year now.

I'm planning to take a last look at it at the [Perl Toolchain Summit](https://act.qa-hackathon.org/pts2019/), expecting to merge it then.

In the mean time, cleaning up the history a bit would make my life more pleasant here.

rivy commented:
# 🧹
* done
